pkg/assembler/graphql/schema/source.graphql

```diff
@@ -0,0 +1,104 @@
+#
+# Copyright 2023 The GUAC Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# NOTE: This is experimental and might change in the future!
+
+# Defines a GraphQL schema for the source trie/tree. This tree is a derivative of
+# the pURL specification where it has a type, namespace, name and finally a qualifier that
+# contain the tag or commit. 
+
+"""
+Source represents a source.
+
+This can be the version control system that is being used.
+
+This node is a singleton: backends guarantee that there is exactly one node with
+the same `type` value.
+
+Also note that this is named `Source`, not `SourceType`. This is only to make
+queries more readable.
+"""
+type Source {
+  type: String!
+  namespaces: [SourceNamespace!]!
+}
+
+"""
+SourceNamespace is a namespace for sources.
+
+This can be represented as the location of the repo (such as github/gitlab/bitbucket)
+
+Namespaces are optional and type specific. Because they are optional, we use
+empty string to denote missing namespaces.
+"""
+type SourceNamespace {
+  namespace: String!
+  names: [SourceName!]!
+}
+
+"""
+SourceName is a url of the repository.
+
+SourceName is mandatory.
+
+This is the first node in the trie that can be referred to by other parts of
+GUAC.
+"""
+type SourceName {
+  name: String!
+  qualifiers: [SourceQualifier!]!
+}
+
+"""
+SourceQualifier containers the commit or tag.
+
+Either a tag or commit needs to be specified.
+
+This node can be referred to by other parts of GUAC.
+"""
+type SourceQualifier {
+  tag: String!
+  commit: String!
+}
+
+"""
+PkgSpec allows filtering the list of packages to return.
+
+Each field matches a qualifier from pURL. Use `null` to match on all values at
+that level. For example, to get all packages in GUAC backend, use a PkgSpec
+where every field is `null`.
+
+Empty string at a field means matching with the empty string.
+"""
+input SourceSpec {
+  type: String
+  namespace: String
+  name: String
+  qualifier: SourceQualifierInput
+}
+
+"""
+SourceQualifierInput is the same as SourceQualifier, but usable as query
+input.
+"""
+input SourceQualifierInput {
+  tag: String
+  commit: String
+}
```

Comment:
I'm thinking to have the spec contain both the `tag` and `commit` fields, making it an error to have both. This way, the query is shorter and more readable: `source(sourceSpec: {tag:"v1.12"})` instead of `source(sourceSpec: {qualifier: {tag:"v1.12"}})` (we still need to error on qualifier if both `tag` and `commit` are given).

Reply:
Now that I think about it more, does it matter if its a tag or commit? Can just be a string stored along with the `sourceName` node.

Reply:
It depends if we want to push processing to upstream or to backend. If caller needs to interpret them, then it can be just a string and the query will have to be `source(sourceSpec: {comittish:"tag=v1.12"")`.

Reply:
I was thinking more like `source(sourceSpec: {comittish:"v1.12"")` or source(sourceSpec: {comittish:"fcba958b73e27cad8b5c8655d46439984d27853b"")

Reply:
I think it is possible to create a tag that looks like a commit, so there could be a source of confusion.

Even GitHub has https://github.com/guacsec/guac/commit/1f083269b1deec24d0a355f9b5c9b2f3e591906c vs https://github.com/guacsec/guac/tree/v0.0.1 (`commit/<commit>` vs `tree/<tag>`)

Reply:
ah good point.

Reply:
I'd still say we can merge these into the spec to simplify query and resolver a little bit more.

Reply:
Discussed offline, this change query semantics so we won't do it.